### PR TITLE
Implement inflation-related functions

### PR DESCRIPTION
**Pull request by AaronVanGeffen**
This is working towards implementing #361.

### CODE REVIEW

File: src/openloco/scenario.cpp [hunk 1]
```diff
@@ -1,12 +1,55 @@
...
 #include "interop/interop.hpp"
 #include "ui/WindowManager.h"
+#include "ui/WindowType.h"
 
 using namespace openloco::interop;
 
 namespace openloco::scenario
 {
+    // Inflation rates?
+    static uint32_t _4FDEC0[32] = {
```
Comment (AaronVanGeffen) [inline src/openloco/scenario.cpp]:
```suggestion
    static const uint32_t _4FDEC0[32] = {
```

File: src/openloco/scenario.cpp [hunk 2]
```diff
@@ -23,4 +66,95 @@ namespace openloco::scenario
...
+    // 0x0046E2C0
+    static void sub_46E2C0(uint16_t year)
+    {
+        for (uint8_t i = 0; i < 32; i++)
+        {
+            currencyMultiplicationFactor[i] = 1024;
+            _525EDE[i] = 0;
+        }
+
+        uint32_t ecx = year - 1900;
```
Comment (AaronVanGeffen) [inline src/openloco/scenario.cpp]: If we change this year to below 1900, the game's balance will be thrown off — everything will become more expensive due to 100 extra years worth of inflation.

Perhaps we should branch off here and not compute any inflation prior to 1900, for now?

Comment (AaronVanGeffen) [reply src/openloco/scenario.cpp]: I've implemented the proposed solution, for now.

**Comment by AaronVanGeffen** [comment]
@kr4g3n This should be it. If you have time, please check out [the AppVeyor CI build](https://ci.appveyor.com/project/AaronVanGeffen/openloco/builds/25205382/artifacts).

**Comment by kr4g3n** [comment]
Thank you. I will try it tomorrow (left my laptop at home today)

**Comment by kr4g3n** [comment]
So I ran this version using Thrillville109's Locomtion Expanded Mod to test the basics.
Without the patch installed, and starting one of Thrillville109's scenarios (start year 1833) I had a starting fund of $18,000,000 and track cost around $125,000.


With the patch installed, I started off with $10,000 and track was around $125/section to lay.

Thrillville109's mod has road vehicles that start in 18000, but the first train appears in 1825. However, [Build Track] option is not disabled. I tried to build track in 1823 and the game crashed to desktop, building in 1825 (after trains appeared) which laid tracks just fine.

So I guess if there was a way to add a control that disable's track laying until a train is discovered, that would be great. Obviously, the first train year would be a variable, based on other's mods,

**Comment by AaronVanGeffen** [comment]
Thanks for testing it out. I wonder what causes it to crash, though — I'm not able to reproduce it.

Screenshot_20190612_111431

I'll it again once I've gotten hold of the mod you've been using.

**Comment by kr4g3n** [comment]
It might be an issue in that mod then, I could never even get it to start on my steam downloaded version (would always freeze at generating terrain) but the mod loads fine in openloco. I'll try and get you a screenshot.

**Comment by kr4g3n** [comment]
I just tried it on the laptop I take to work, no issues. Trying again on my Alienware M18Xr2 incase it's machine dependant...

**Comment by kr4g3n** [comment]
OK tried on my Alienware - first I retried the scenario I built last night. SAME THING. 

This is the scenario I built. Just in case you can find anything  in there? I set it to flat and changed the sea level, hill density, no of forests, no of towns, and number of industries.
[TrainTable.zip](https://github.com/OpenLoco/OpenLoco/files/3281775/TrainTable.zip)

I also seleceted ALL vehicles and Airports from thrillville's mod.
[LocomotionExpanded.zip](https://github.com/OpenLoco/OpenLoco/files/3281781/LocomotionExpanded.zip)

I weill have to do some more testing later

**Comment by kr4g3n** [comment]
First, sorry, but I uploaded the wrong scenario file earlier, I'm at work right now so can't upload the other one.

I have however tried some more tests using the boulder breakers map, each time I was able to place tracks etc, but it randomly crashes to desktop. 

Is there a dump file I can send your way? 

Both my machines are Win10

In addition, I noticed that if I start any date other than 1800, I don't actually get that start date. 

Example one game I started in 1823, but the actual start year was 1843.
And if I tried to start in 1825, it started in 1849.

File: src/openloco/scenario.cpp [hunk 3]
```diff
@@ -23,4 +66,97 @@ namespace openloco::scenario
     {
         call(0x0043C90C);
     }
+
+    // 0x0046E239
+    // NB: called in sub_46E2C0 below, as well in openloco::date_tick.
+    void sub_46E239()
+    {
+        for (uint8_t i = 0; i < 32; i++)
```
Comment (marijnvdwerf) [inline src/openloco/scenario.cpp]: I think the entire highlighted area is useless, as `_525EDE` is only ever filled with zeroes.
loc_46e238

File: src/openloco/scenario.cpp [hunk 4]
```diff
@@ -1,12 +1,55 @@
...
+        20,
+        20,
+        20,
+        20,
+        20,
+    };
+
+    static loco_global<uint32_t[32], 0x00525E5E> currencyMultiplicationFactor;
+
+    // NB: _525EDE[2] is dword_525E66, referred to by four subs for (price) computation.
```
Comment (marijnvdwerf) [inline src/openloco/scenario.cpp]: Comment is incorrect, as it applies to the previous variable instead

File: src/openloco/scenario.cpp [hunk 5]
```diff
@@ -23,4 +66,97 @@ namespace openloco::scenario
...
+    {
+        for (uint8_t i = 0; i < 32; i++)
+        {
+            currencyMultiplicationFactor[i] = 1024;
+            _525EDE[i] = 0;
+        }
+
+        // OpenLoco allows 1800 as the minimum year, whereas Locomotion uses 1900.
+        // Treat years before 1900 as though they were 1900 to not change vanilla scenarios.
+        uint32_t ecx = std::max<uint32_t>(year, 1900) - 1900;
```
Comment (marijnvdwerf) [inline src/openloco/scenario.cpp]: something like `std::clamp(year, 1900, 2030) - 1900` would be easier to read.

File: src/openloco/scenario.cpp [hunk 6]
```diff
@@ -23,4 +66,97 @@ namespace openloco::scenario
...
+        _scenario_ticks = 0;
+
+        loco_global<uint16_t, 0x0052622E> _52622E;
+        _52622E = 0;
+
+        loco_global<uint8_t, 0x00525FB5> _525FB5;
+        _525FB5 = 1;
+
+        // NB: only called here.
+        call(0x0042FDE2);
```
Comment (marijnvdwerf) [inline src/openloco/scenario.cpp]: Goes through all companies and sets the available vehicles.

File: src/openloco/scenario.cpp (hunk 3 above)
Comment (AaronVanGeffen) [reply src/openloco/scenario.cpp]: What makes you sure it's zeros? I think it represents the highest 16 bits of a `money48` struct.

File: src/openloco/scenario.cpp [hunk 7]
```diff
@@ -1,12 +1,55 @@
...
 #include "interop/interop.hpp"
 #include "ui/WindowManager.h"
+#include "ui/WindowType.h"
 
 using namespace openloco::interop;
 
 namespace openloco::scenario
 {
+    // Inflation rates?
+    static const uint32_t _4FDEC0[32] = {
```
Comment (marijnvdwerf) [inline src/openloco/scenario.cpp]: List of fixed-point percentages. (12 bits for fractions)

File: src/openloco/scenario.cpp [hunk 8]
```diff
@@ -1,12 +1,55 @@
...
+        20,
+        20,
+        20,
+        20,
+        20,
+        20,
+        20,
+    };
+
+    static loco_global<uint32_t[32], 0x00525E5E> currencyMultiplicationFactor;
```
Comment (marijnvdwerf) [inline src/openloco/scenario.cpp]: Assuming these values start at `1`, this is fixed point with 10 fractional bits.

File: src/openloco/scenario.cpp (hunk 3 above)
Comment (marijnvdwerf) [reply src/openloco/scenario.cpp]: As I said, `525EDE` is only ever filled with zeroes. Just look at the references.

**Comment by kr4g3n** [comment]
Is there an updated build to test?

**Comment by AaronVanGeffen** [comment]
@marijnvdwerf Thanks for the review comments.

> Is there an updated build to test?

@kr4g3n No, there have not been any changes since the last build yet. I've been a bit too busy, unfortunately. Check back soon. :)

**Comment by kr4g3n** [comment]
No worries, I appreciate that you're even looking into it.

I know I've asked this before, but when the game crashes, does it put a crash dump file anywhere?

**Comment by AaronVanGeffen** [comment]
It does not, yet. You could try launching the game from a command prompt, however. Slim chance, but there might be an error printed to the console.

**Comment by marijnvdwerf** [comment]
Transport Tycoon mobile seems to have this in a separate `Money` namespace.
```
cTTE_Handler_Money::~cTTE_Handler_Money()
cTTE_Handler_Money::~cTTE_Handler_Money()
cTTE_Handler_Money::Clear()
cTTE_Handler_Money::cTTE_Handler_Money()
cTTE_Handler_Money::cTTE_Handler_Money()
cTTE_Handler_Money::MonthlyInflation()
cTTE_Handler_Money::CalculateInflationToBaseYear(int, int, int)
cTTE_Handler_Money::GetGeneralCostReferenceWholeValue()
cTTE_Handler_Money::CalcCost(long, int, int)
cTTE_Handler_Money::CalcCostWithUnitsDistanceAndTime(long, int, int, int, int, int)
cTTE_Handler_Money::CalcCostWithUnitsDistanceAndTimeWithFractional(long, int, int, int, int, int)
cTTE_Handler_Money::GetLocalisedCurrencyValue(long, int)
cTTE_Handler_Money::GetLocalisedCurrencyValue(long long, int)
cTTE_Handler_Money::GetLocalisedCurrencyString(long, int, char*, int)
cTTE_Handler_Money::GetLocalisedCurrencyString(long long, int, char*, int)
```

**Comment by marijnvdwerf** [comment]
Also, I think this is a good moment to add a templated `fixed_point` type.

**Comment by AaronVanGeffen** [comment]
@marijnvdwerf I'm not sure what you're getting at with these notions of fixed points. Rather than going back and forth, would you mind implementing the desired changes yourself?

**Comment by AaronVanGeffen** [comment]
I have rebased the PR to account for the recent wave of PRs that renamed functions to use camelCase. In the process, I have dropped @marijnvdwerf's commits that introduced a work-in-progress 'fixed point' class. This class was left unfinished, and unfortunately caused this PR to stall for more than a year.

Going forward, I hope to finish the functions can be corrected in a way that makes them correct. However, I think refactoring them to a fixed point class should really be done in another PR.

File: src/OpenLoco/Scenario.cpp [hunk 9]
```diff
@@ -12,6 +14,7 @@
 #include "Title.h"
 #include "TownManager.h"
 #include "Ui/WindowManager.h"
+#include "Ui/WindowType.h"
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Scenario.cpp:17]: No longer needed

File: src/OpenLoco/Scenario.cpp [hunk 10]
```diff
@@ -142,9 +145,41 @@ namespace OpenLoco::Scenario
...
+        _current_day = eax;
+
+        loco_global<uint16_t, 0x00525E30> _day_progression;
+        _day_progression = 0;
+
+        loco_global<int8_t, 0x00525E34> _current_month;
+        _current_month = 0;
+
+        loco_global<int8_t, 0x00525E35> _current_day_of_month;
+        _current_day_of_month = 0;
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Scenario.cpp:168]: Use `Date::setDate`

File: src/OpenLoco/Scenario.cpp [hunk 11]
```diff
@@ -142,9 +145,41 @@ namespace OpenLoco::Scenario
     // 0x0049685C
     void initialiseDate(uint16_t year)
     {
-        registers regs;
-        regs.ax = year;
-        call(0x0049685C, regs);
+        loco_global<int16_t, 0x00525E32> _current_year;
+        _current_year = year;
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Scenario.cpp:149]: `Date::getCurrentYear`

File: src/OpenLoco/Economy.cpp [hunk 12]
```diff
@@ -1,21 +1,101 @@
...
+        {
+            currencyMultiplicationFactor[i] = 1024;
+            _525EDE[i] = 0;
+        }
+
+        // OpenLoco allows 1800 as the minimum year, whereas Locomotion uses 1900.
+        // Treat years before 1900 as though they were 1900 to not change vanilla scenarios.
+        uint32_t yearCount = std::clamp<uint32_t>(year, 1900, 2030) - 1900;
+
+        yearCount *= 12;
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Economy.cpp:94]: Initialise in one assignment as `monthCount`?

**Comment by AaronVanGeffen** [comment]
The scenario stuff has been split off to #791 and merged to master, leaving only the inflation routines here. Help is welcome.

File: src/OpenLoco/Economy.cpp [hunk 13]
```diff
@@ -1,21 +1,101 @@
...
+        {
+            uint32_t ebx = currencyMultiplicationFactor[i];
+
+            // Inflating prices?
+            uint64_t edx_eax = _525EDE[i] * _4FDEC0[i];
+            edx_eax >>= 12;
+
+            // Storing two halves of the money struct?
+            _525EDE[i] += static_cast<uint32_t>(edx_eax);
+            currencyMultiplicationFactor[i] += static_cast<uint32_t>(edx_eax >> 32);
```
Comment (duncanspumpkin) [inline src/OpenLoco/Economy.cpp:65]: this isn't quite correct. There is a carry from 525ede.
You'll need to do:
```
uint64_t new525 = _525EDE[i] + static_cast<uint32_t>(edx_eax);
_525EDE[i] = static_cast<uint32_t>(new525);
currencyMultiplicationFactor[i] += static_cast<uint32_t>(edx_eax >> 32) + (new525 >> 32);
```

File: src/OpenLoco/Economy.cpp [hunk 14]
```diff
@@ -1,21 +1,101 @@
...
+
+            // Inflating prices more?
+            edx_eax = ebx * _4FDEC0[i];
+            edx_eax >>= 12;
+            currencyMultiplicationFactor[i] += static_cast<uint32_t>(edx_eax);
+        }
+
+        call(0x004375F7);
```
Comment (duncanspumpkin) [inline src/OpenLoco/Economy.cpp:73]: Turn this into a sub function stub. Does some calcs on cargos.

File: src/OpenLoco/Economy.cpp (hunk 13 above)
Comment (duncanspumpkin) [reply src/OpenLoco/Economy.cpp:65]: Wait, there is a much easier way of doing this its just simple maths.

```
uint64_t existingInflation = _525EDE[i] + (currencyMultiplicationFactor[i] << 32);
uint64_t additonalInflation = (existingInflation *  _4FDEC0[i]) / 4096;
uint64_t newInflation = existingInflation + additionalInflation;
_525EDE[i] = static_cast<uint32_t>(newInflation);
currencyMultiplicationFactor[i] = static_cast<uint32_t>(newInflation>> 32);
```

Comment (duncanspumpkin) [reply src/OpenLoco/Economy.cpp:65]: Oh wait but CS didn't do his calculation correctly. So this is wrong.

Comment (duncanspumpkin) [reply src/OpenLoco/Economy.cpp:65]:
```
        for (uint8_t i = 0; i < 32; i++)
        {
            currencyMultiplicationFactor[i] += (static_cast<uint64_t>(_4FDEC0[i]) * currencyMultiplicationFactor[i]) >> 12;
        }
```

File: src/openloco/scenario.cpp (hunk 3 above)
Comment (duncanspumpkin) [reply src/openloco/scenario.cpp]: You can see what CS tried to do but he performed the calc wrong. The lower bits from the second shrd were meant to be put into 525EDE but instead are discarded. So 525EDE does indeed do nothing.

File: src/OpenLoco/Economy.cpp [hunk 15]
```diff
@@ -1,21 +1,90 @@
...
 #include "Interop/Interop.hpp"
+#include "Ui/WindowManager.h"
+#include "Ui/WindowType.h"
 
 using namespace OpenLoco::Interop;
 
 namespace OpenLoco::Economy
 {
+    // Inflation rates?
+    static const uint32_t _4FDEC0[32] = {
```
Comment (AaronVanGeffen) [inline src/OpenLoco/Economy.cpp:11]: Shall we name this array then?
```suggestion
    static const uint32_t _inflationFactors[32] = {
```

File: src/OpenLoco/Economy.cpp (hunk 13 above)
Comment (AaronVanGeffen) [reply src/OpenLoco/Economy.cpp]: Wow. I'm pleasantly surprised that chunk of code just boiled down to a oneliner in the end. Thank you very much!

**Comment by AaronVanGeffen** [comment]
For what it's worth, I've played the game a few hours on this branch tonight. I'm fairly sure the inflation is working properly.

**Review by duncanspumpkin — APPROVED** [review]

**Pull request merged**